src/Microsoft.AspNetCore.Mvc.Analyzers/SymbolApiResponseMetadataProvider.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers
+{
+    internal class SymbolApiResponseMetadataProvider
+    {
+        internal static IList<ApiResponseMetadata> GetResponseMetadata(ApiControllerTypeCache typeCache, IMethodSymbol methodSymbol)
```

Comment:
> // Internal for unit testing

So we generally stopped using that since we use `internal` for sharing code inside an assembly as *intended*. 

The code's sort of structured like so: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Analyzers/AttributesShouldNotBeAppliedToPageModelAnalyzer.cs#L26

1) We populate the cache
2) Verify enough of the symbols we need are available
3) Do the work. Doing the work would involve calling to this type. 

This type is sort of meant to be a utility type which has the same shape as https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.ApiExplorer/ApiResponseTypeProvider.cs#L33. I did start off doing something similar earlier -https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Analyzers/ViewFeaturesAnalyzerContext.cs#L12 - but it feels weird to have utility methods on the context \ cache type.

Reply:
Just out of curiosity.  What does having a separate cache type buy us? Are the lifetimes of the analyzers typically super short lived?

Reply:
The analyzer itself is long-lived but I wanted to avoid the work of symbol lookup per invocation of a symbol action. Roslyn uses a similar pattern in their analyzers too - https://github.com/dotnet/roslyn-analyzers/blob/26fe065161a871718c717da1821d82c4decbc8e9/src/Microsoft.CodeAnalysis.Analyzers/CSharp/CSharpImmutableObjectMethodAnalyzer.cs#L53-L66